scripts/await-previous-builds.js

```diff
@@ -0,0 +1,73 @@
+// @ts-check
+// basic idea came from https://github.com/bellkev/circle-lock-test
+
+const fetch = require("node-fetch").default
+const chalk = require("chalk").default
+const { throttle } = require("lodash")
+
+/**
+ * Returns a promise that resolves once all previous circleCI builds on the current branch
+ * have finished (or failed for whatever reason).
+ * @param {{pollInterval?: number, logInterval?:number}?} props
+ */
+async function awaitPreviousBuilds({ pollInterval = 5000, logInterval = 30000 } = {}) {
+  const { CIRCLE_TOKEN, CIRCLE_BRANCH, CIRCLECI } = process.env
+
+  if (CIRCLECI !== "true") {
+    throw new Error("awaitPreviousBuilds only works on CI")
+  }
+
+  if (!CIRCLE_TOKEN) {
+    throw new Error(
+      "Please add CIRCLE_TOKEN in env. Must have read access for the project. See https://circleci.com/docs/api/#authentication (make sure it's scope: all)"
+    )
+  }
+
+  const notifyAboutWaiting = throttle(({ numActiveBuilds }) => {
+    console.log(chalk.cyan(`⏱ Waiting for ${numActiveBuilds} previous build(s) on ${CIRCLE_BRANCH} to finish...`))
+  }, logInterval)
+
+  while (true) {
+    const numActiveBuilds = await getNumberOfPreviousActiveBuilds()
+
+    if (numActiveBuilds === 0) {
+      console.log(chalk.cyan(`✨ Previous builds on ${CIRCLE_BRANCH} have all finished!`))
+      return
+    }
+
+    notifyAboutWaiting({ numActiveBuilds })
+
+    await new Promise(r => setTimeout(r, pollInterval))
+  }
+}
+
+async function getNumberOfPreviousActiveBuilds() {
+  const {
+    CIRCLE_BRANCH,
+    CIRCLE_BUILD_NUM,
+    CIRCLE_PROJECT_REPONAME,
+    CIRCLE_PROJECT_USERNAME,
+    CIRCLE_TOKEN,
+  } = process.env
+
+  const res = await fetch(
+    `https://circleci.com/api/v1/project/${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}?circle-token=${CIRCLE_TOKEN}&limit=100`,
+    { headers: { Accept: "application/json" } }
+  )
```

Comment:
Similarly to the comment below, how feasible would it be for exceptions to be thrown from here for temporary issues, in which case we might prefer a retry over a failure as it would fail the release build?

Reply:
I think it's no biggie. The developer can manually restart the build if that situation arises, which seems unlikely. Since this runs before any changes are made to the repo, on failures it won't be left in a bad state and manual retries should Just Work™

Reply:
Hmm, right, I was thinking of Eigen builds where the dev having to restart a release build actually can be a biggie (due to deadlines/time-constraints).